File: src/main/java/com/github/yufiriamazenta/deathmsg/commands/FilterDeathMessageCmd.java
package com.github.yufiriamazenta.deathmsg.commands;

import com.github.yufiriamazenta.deathmsg.DeathMessage;
import com.github.yufiriamazenta.deathmsg.util.LangUtil;
import crypticlib.command.IPluginCommand;
import crypticlib.command.ISubCommand;
import org.bukkit.NamespacedKey;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.plugin.Plugin;

import java.util.*;

public class FilterDeathMessageCmd implements IPluginCommand {

    private static final NamespacedKey filterKey = new NamespacedKey(DeathMessage.INSTANCE, "death_msg_filter");

    @Override
    public boolean onCommand(CommandSender commandSender, Command command, String s, String[] strings) {
        if (!(commandSender instanceof Player player)) {
            LangUtil.msg(commandSender, "plugin_message.only_player");
            return true;
        }
        if (strings.length == 0) {
            toggleFilter(player);
        } else {
            switch (strings[0]) {
                case "on" -> setFilterOn(player);
                case "off" -> setFilterOff(player);
                default -> toggleFilter(player);
            }
        }
        return true;
    }

    public void toggleFilter(Player player) {
        PersistentDataContainer dataContainer = player.getPersistentDataContainer();
        Byte val = dataContainer.get(filterKey, PersistentDataType.BYTE);
        if (val == null) {
            setFilterOn(player);
            return;
        }
        if (val == 0) {
            setFilterOn(player);
        } else {
            setFilterOff(player);
        }
    }

    public void setFilterOn(Player player) {
        PersistentDataContainer dataContainer = player.getPersistentDataContainer();
        dataContainer.set(filterKey, PersistentDataType.BYTE, (byte)1);
        LangUtil.msg(player, "plugin_message.filter_on");
    }

    public void setFilterOff(Player player) {
        PersistentDataContainer dataContainer = player.getPersistentDataContainer();
        dataContainer.set(filterKey, PersistentDataType.BYTE, (byte)0);
        LangUtil.msg(player, "plugin_message.filter_off");
    }

    public static NamespacedKey getFilterKey() {
        return filterKey;
    }

    @Override
    public Plugin getPlugin() {
        return DeathMessage.getInstance();
    }

    @Override
    public List<String> onTabComplete(CommandSender commandSender, Command command, String s, String[] strings) {
        switch (strings.length) {
            case 0, 1 -> {
                List<String> list = new ArrayList<>(List.of("off", "on"));
                list.removeIf(str -> !str.startsWith(strings[0]));
                return list;
            }default -> {
                return Collections.singletonList("");
            }
        }
    }

    @Override
    public Map<String, ISubCommand> subCommands() {
        return new HashMap<>();
    }
}

File: src/main/kotlin/com/github/yufiriamazenta/deathmsg/commands/FilterDeathMessageCmd.kt
package com.github.yufiriamazenta.deathmsg.commands

import com.github.yufiriamazenta.deathmsg.DEATH_MESSAGE
import com.github.yufiriamazenta.deathmsg.util.LangUtil
import crypticlib.command.CommandHandler
import crypticlib.command.CommandInfo
import crypticlib.command.annotation.Command
import crypticlib.perm.PermInfo
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.persistence.PersistentDataType

@Command
object FilterDeathMessageCmd : CommandHandler(CommandInfo("deathmessagefilter", PermInfo("deathmessage.command.filter"), arrayOf("dmf"))) {

    override fun execute(sender: CommandSender, args: MutableList<String>): Boolean {
        if (sender !is Player) {
            LangUtil.sendLang(sender, "plugin_message.only_player")
            return true
        }
        if (args.isEmpty()) {
            toggleFilter(sender)
        } else {
            when (args[0]) {
                "on" -> setFilterOn(sender)
                "off" -> setFilterOff(sender)
                else -> toggleFilter(sender)
            }
        }
        return true
    }

    override fun tab(sender: CommandSender, args: MutableList<String>): MutableList<String> {
        return when (args.size) {
            0, 1 -> {
                val list: MutableList<String> = mutableListOf("off", "on")
                list.removeIf { str: String -> !str.startsWith(args[0]) }
                list
            }

            else -> {
                mutableListOf("")
            }
        }
    }

    private fun toggleFilter(player: Player) {
        val filterFlag = DEATH_MESSAGE.isPlayerDeathMsgFilterOn(player)
        if (filterFlag) {
            setFilterOff(player)
        } else {
            setFilterOn(player)
        }
    }

    private fun setFilterOn(player: Player) {
        val dataContainer = player.persistentDataContainer
        dataContainer[DEATH_MESSAGE.deathMessageFilterKey, PersistentDataType.BYTE] = 1.toByte()
        LangUtil.sendLang(player, "plugin_message.filter_on")
    }

    private fun setFilterOff(player: Player) {
        val dataContainer = player.persistentDataContainer
        dataContainer[DEATH_MESSAGE.deathMessageFilterKey, PersistentDataType.BYTE] = 0.toByte()
        LangUtil.sendLang(player, "plugin_message.filter_off")
    }

}
